src/rust/ide/enso-protocol/src/project_manager.rs

```diff
@@ -0,0 +1,382 @@
+use json_rpc::make_param_map;
+use json_rpc::make_arg;
+use futures::Stream;
+use serde::Serialize;
+use serde::Deserialize;
+use std::future::Future;
+use uuid::Uuid;
+
+
+
+// =============
+// === Event ===
+// =============
+
+/// Event emitted by the Project Manager `Client`.
+pub type Event = json_rpc::handler::Event<Notification>;
+
+
+
+// ====================
+// === Notification ===
+// ====================
+
+/// Notification generated by the Project Manager.
+#[derive(Clone,Copy,Debug,PartialEq)]
+#[derive(Serialize, Deserialize)]
+#[serde(tag="method", content="params")]
+pub enum Notification {}
+
+
+
+// ===================
+// === RPC Methods ===
+// ===================
+
+make_rpc_methods! {
+/// An interface containing all the available project management operations.
+trait API {
+    /// Requests that the project picker open a specified project. This operation also
+    /// includes spawning an instance of the language server open on the specified project.
+    #[MethodInput=OpenProjectInput,rpc_name="project/open",result=open_project_result,set_result=set_open_project_result]
+    fn open_project(&self, project_id:Uuid) -> OpenProjectResponse;
+
+    /// Requests that the project picker close a specified project. This operation
+    /// includes shutting down the language server gracefully so that it can persist state to disk
+    /// as needed.
+    #[MethodInput=CloseProjectInput,rpc_name="project/close",result=close_project_result,
+    set_result=set_close_project_result]
+    fn close_project(&self, project_id:Uuid) -> ();
+
+    /// Requests that the project picker lists the user's most recently opened
+    /// projects.
+    #[MethodInput=ListRecentInput,rpc_name="project/listRecent",result=list_recent_result,
+    set_result=set_list_recent_result]
+    fn list_recent(&self, number_of_projects:u32) -> ProjectListResponse;
+
+    /// Requests the creation of a new project.
+    #[MethodInput=CreateProjectInput,rpc_name="project/create",result=create_project_result,set_result=set_create_project_result]
+    fn create_project(&self, name:String) -> CreateProjectResponse;
+
+    /// Requests the deletion of a project.
+    #[MethodInput=DeleteProjectInput,rpc_name="project/delete",result=delete_project_result,set_result=set_delete_project_result]
+    fn delete_project(&self, project_id:Uuid) -> ();
+
+    /// Requests a list of sample projects that are available to the user.
+    #[MethodInput=ListSampleInput,rpc_name="project/listSample",result=list_sample_result,set_result=set_list_sample_result]
+    fn list_sample(&self, num_projects:u32) -> ProjectListResponse;
+}
+}
+
+/// IP address with host and port.
+#[derive(Debug,Clone,Serialize,Deserialize,PartialEq)]
+pub struct IpWithSocket {
+    host : String,
+    port : u16
+}
+
+/// Project information, such as name, its id and last time it was opened.
+#[derive(Debug,Clone,Serialize,Deserialize,PartialEq)]
+pub struct ProjectMetaData {
+    name        : String,
```

Comment:
this should be newtype.

Reply:
Why?

Reply:
I'm not sure. I would create a newtype if we actually will have many names and start to mix them together

Reply:
Making newtype would not hurt anyone and I highly doubt we will ever do it in the future if we leave it this way now ;)

Reply:
O, actually it hurts me a lot, when I must handle with create/access/modify functions, and when you shrinkwrap the code completion stops working.